### PR TITLE
Added site runtime preview code behind ECS Config

### PR DESCRIPTION
**Pull request by ashwani123p**
#### Overview

This PR introduces a command `Preview Site` that enables launching Microsoft Edge DevTools directly within Visual Studio Code for the PreviewSite functionality within VSCode Desktop. On executing the command, the Microsoft Edge Tools extension will be triggered to provide an in-editor preview environment.

#### Changes Made

- **launchBrowserAndDevToolsWithinVsCode** in PreviewSite: Added a new command to launch the Edge DevTools project within VS Code.
-  ECS Flag for Feature Toggle (enableSiteRuntimePreview): Added an ECS configuration setting to determine feature availability.
- Changes in extension.ts file are:
   - When ECS config is enabled, website details were fetched from PPAPI Service, and the site preview launched with the correct URL which matched with websiteRecordId in site repo.

### CODE REVIEW

**Comment by ashwani123p** [comment]
> @ashwani123p please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

File: src/common/services/PPAPIService.ts [hunk 1]
```diff
@@ -33,6 +33,37 @@ export class PPAPIService {
...
+                headers: getCommonHeaders(accessToken)
+            });
+
+            if (response.ok) {
+                const websiteDetailsArray = await response.json() as unknown as IWebsiteDetails[];
+                return websiteDetailsArray;
+            }
+        }
+        catch (error) {
+            sendTelemetryEvent(telemetry, { eventName: VSCODE_EXTENSION_GET_CROSS_GEO_DATA_MOVEMENT_ENABLED_FLAG_FAILED, errorMsg: (error as Error).message });
```
Comment (amitjoshi438) [inline src/common/services/PPAPIService.ts:62]: Is the telemetry event name correct?

File: src/common/utilities/WorkspaceInfoFinderUtil.ts [hunk 2]
```diff
@@ -36,3 +39,24 @@ export function getPortalsOrgURLs(workspaceRootFolders: WorkspaceFolder[] | null
...
+        const websiteYmlPath = path.join(workspaceRootFolder.uri, 'website.yml');
+        if (fs.existsSync(websiteYmlPath)) {
+            const fileContent = fs.readFileSync(websiteYmlPath, 'utf8');
+            const parsedYaml = parse(fileContent);
+            if (parsedYaml && parsedYaml.adx_websiteid) {
+                return parsedYaml.adx_websiteid;
+            }
+        }
+    } catch (exception) {
+        sendTelemetryEvent(telemetry, { methodName: getWebsiteRecordID.name, eventName: 'getWebsiteRecordID', exception: exception as Error });
```
Comment (amitjoshi438) [inline src/common/utilities/WorkspaceInfoFinderUtil.ts:59]: Can we move this string to constant file?

File: src/client/runtimeSitePreview/PreviewSite.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                if (vscodeFolderPath && !vscodeFolderExisted && fs.existsSync(vscodeFolderPath)) {
+                    const files = fs.readdirSync(vscodeFolderPath);
+                    if (files.length === 0) {
+                        fs.rmdirSync(vscodeFolderPath);
+                    }
+                }
+            }
+        } else {
+            const install = await vscode.window.showWarningMessage(
+                `The extension "${edgeToolsExtensionId}" is required to run this command. Do you want to install it now?`,
```
Comment (amitjoshi438) [inline src/client/runtimeSitePreview/PreviewSite.ts:62]: If we are showing it to user, localization is required

File: src/client/runtimeSitePreview/LaunchJsonHelper.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+import * as vscode from 'vscode';
+
+export async function updateLaunchJsonConfig(url: string): Promise<void> {
+
+    const workspaceFolders = vscode.workspace.workspaceFolders;
+    if (!workspaceFolders) {
+        vscode.window.showErrorMessage('No workspace folder is open.');
```
Comment (amitjoshi438) [inline src/client/runtimeSitePreview/LaunchJsonHelper.ts:12]: Localization needed

File: src/client/extension.ts [hunk 5]
```diff
@@ -270,6 +263,47 @@ export async function activate(
         vscode.commands.executeCommand('setContext', 'powerpages.websiteYmlExists', false);
     }
 
+    const registerPreviewShowCommand = async () => {
```
Comment (amitjoshi438) [inline src/client/extension.ts:266]: Can we move the preview related code to a separate file?

File: src/client/extension.ts [hunk 6]
```diff
@@ -299,6 +333,23 @@ export async function deactivate(): Promise<void> {
     disposeNotificationPanel();
 }
```
Comment (amitjoshi438) [inline src/client/extension.ts:335]: This also we can move to a separate file along with above code.

File: src/client/extension.ts [hunk 7]
```diff
@@ -270,6 +263,47 @@ export async function activate(
...
+                            commandId: "microsoft-powerapps-portals.preview-show"
+                        });
+                        PortalWebView.createOrShow();
+                    }
+                )
+            );
+        } else {
+            _context.subscriptions.push(
+                vscode.commands.registerCommand(
+                    "microsoft-powerapps-portals.preview-show",
```
Comment (amitjoshi438) [inline src/client/extension.ts:297]: Should we add telemetry here also, like above command?

File: src/client/runtimeSitePreview/PreviewSite.ts (hunk 3 above)
Comment (ashwani123p) [reply src/client/runtimeSitePreview/PreviewSite.ts:62]: sure

File: src/common/services/PPAPIService.ts (hunk 1 above)
Comment (ashwani123p) [reply src/common/services/PPAPIService.ts]: corrected

File: src/common/utilities/WorkspaceInfoFinderUtil.ts (hunk 2 above)
Comment (ashwani123p) [reply src/common/utilities/WorkspaceInfoFinderUtil.ts]: done

File: src/client/extension.ts (hunk 5 above)
Comment (ashwani123p) [reply src/client/extension.ts:266]: this is registering command so extension.js is right location for it, moved the other code of preview site to separate file

File: src/client/extension.ts (hunk 6 above)
Comment (ashwani123p) [reply src/client/extension.ts]: done

File: src/client/extension.ts (hunk 7 above)
Comment (ashwani123p) [reply src/client/extension.ts:297]: added a common telemetry irrelevance of enableSiteRuntimePreview config enabled or disabled

File: src/client/runtimeSitePreview/PreviewSite.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                        fs.rmdirSync(vscodeFolderPath);
+                    }
+                }
+            }
+        } else {
+            const install = await vscode.window.showWarningMessage(
+                vscode.l10n.t(
+                `The extension Microsoft Edge Tools is required to run this command. Do you want to install it now?`,
+                'Install', 'Cancel'
+            ));
```
Comment (priyanshu92) [inline src/client/runtimeSitePreview/PreviewSite.ts:96]: Is this working or do we need to pass `Install` and `Cancel` as arguments to `showWarningMessage`?

File: src/client/runtimeSitePreview/PreviewSite.ts
[the remark this answers is not in the document]
Comment (ashwani123p) [reply src/client/runtimeSitePreview/PreviewSite.ts:1]: done

File: src/client/runtimeSitePreview/PreviewSite.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import { ServiceEndpointCategory } from '../../common/services/Constants';
+import { PPAPIService } from '../../common/services/PPAPIService';
+import { VSCODE_EXTENSION_GET_WEBSITE_RECORD_ID_EMPTY } from '../../common/services/TelemetryConstants';
+
+export class PreviewSite {
+
+    static isSiteRuntimePreviewEnabled(): boolean {
+        const enableSiteRuntimePreview = ECSFeaturesClient.getConfig(EnableSiteRuntimePreview).enableSiteRuntimePreview
+
+        if(enableSiteRuntimePreview === undefined) {
```
[the remark this answers is not in the document]
Comment (ashwani123p) [reply src/client/runtimeSitePreview/PreviewSite.ts]: done

File: src/client/runtimeSitePreview/PreviewSite.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+                if (launchJsonPath && fs.existsSync(launchJsonPath)) {
+                    originalLaunchJsonContent = fs.readFileSync(launchJsonPath, 'utf8');
+                }
+            }
+
+            await updateLaunchJsonConfig(webSitePreviewURL);
+
+            try {
+                // Added a 2-second delay before executing the launchProject command to handle the case where the launch.json file is not saved yet
+                await new Promise(resolve => setTimeout(resolve, 2000));
```
[the remark this answers is not in the document]
Comment (ashwani123p) [reply src/client/runtimeSitePreview/PreviewSite.ts]: Adding a bug https://dynamicscrm.visualstudio.com/OneCRM/_workitems/edit/4469030, to resolve it in phase 2 release

**Review by ashishchoudhary001 — APPROVED** [review]

**Pull request merged**